Add csv support to stats_over_http

ezelkow1 opened the pull request:
<!--
************ STOP!! ************
If this Pull Request is intended to fix a security vulnerability, DO NOT submit it! Instead, contact
the Apache Software Foundation Security Team at security@trafficcontrol.apache.org and follow the
guidelines at https://www.apache.org/security/ regarding vulnerability disclosure.
-->
## What does this PR (Pull Request) do?
This uses the same parsing method as astats csv and provides almost double the performance with less than half the allocations of the json version.

Also add a null check for dsStats. Its possible to do a request to ATS to a DS that does not exist, ATS will then insert stats for that request into the remap stats list. When it gets to TM though it will not be in the list of known DSs so if we try to index by it it will end up null
<!-- Explain the changes you made here. If this fixes an Issue, identify it by
replacing the text in the checkbox item with the Issue number e.g.

- [x] This PR fixes #9001 OR is not related to any Issue

^ This will automatically close Issue number 9001 when the Pull Request is
merged (The '#' is important).

Be sure you check the box properly, see the "The following criteria are ALL
met by this PR" section for details.
-->

- [x] This PR fixes #REPLACE_ME OR is not related to any Issue <!-- You can check for an issue here: https://github.com/apache/trafficcontrol/issues -->


## Which Traffic Control components are affected by this PR?
<!-- Please delete all components from this list that are NOT affected by this
Pull Request. Also, feel free to add the name of a tool or script that is
affected but not on the list.

Additionally, if this Pull Request does NOT affect documentation, please
explain why documentation is not required. -->

- Traffic Monitor

## What is the best way to verify this PR?
<!-- Please include here ALL the steps necessary to test your Pull Request. If
it includes tests (and most should), outline here the steps needed to run the
tests. If not, lay out the manual testing procedure and please explain why
tests are unnecessary for this Pull Request. -->

- Add health.polling.format = stats_over_http to a server profile
- Set http_polling_format for TM to be "text/csv"
- Check if stats are pulled correctly in TM

## If this is a bug fix, what versions of Traffic Control are affected?
<!-- If this PR fixes a bug, please list here all of the affected versions - to
the best of your knowledge. It's also pretty helpful to include a commit hash
of where 'master' is at the time this PR is opened (if it affects master),
because what 'master' means will change over time. For example, if this PR
fixes a bug that's present in master (at commit hash '1df853c8'), in v4.0.0,
and in the current 4.0.1 Release candidate (e.g. RC1), then this list would
look like:

- master (1df853c8)
- 4.0.0
- 4.0.1 (RC1)

If you don't know what other versions might have this bug, AND don't know how
to find the commit hash of 'master', then feel free to leave this section
blank (or, preferably, delete it entirely).
 -->


## The following criteria are ALL met by this PR
<!-- Check the boxes to signify that the associated statement is true. To
"check a box", replace the space inside of the square brackets with an 'x'.
e.g.

- [ x] <- Wrong
- [x ] <- Wrong
- [] <- Wrong
- [*] <- Wrong
- [x] <- Correct!

-->

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)


## Additional Information
<!-- If you would like to include any additional information on the PR for
potential reviewers please put it here.

Some examples of this would be:

- Before and after screenshots/gifs of the Traffic Portal if it is affected
- Links to other dependent Pull Requests
- References to relevant context (e.g. new/updates to dependent libraries,
mailing list records, blueprints)

Feel free to leave this section blank (or, preferably, delete it entirely).
-->
Benchmarks comparing all stats parsing, both astats and stats over http with json and csv. In both CSV instances they are ~100% faster with less than half as many allocations

BenchmarkAstatsJson-8   	    5276	    202209 ns/op	  134453 B/op	    2028 allocs/op
BenchmarkAstatsCSV-8    	    9963	    111798 ns/op	  110001 B/op	     798 allocs/op
BenchmarkStatsJson-8    	    5512	    227104 ns/op	  134991 B/op	    2082 allocs/op
BenchmarkStatsCSV-8     	   10000	    108320 ns/op	  111217 B/op	     837 allocs/op

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

zrhoffman reviewed (changes requested):
Our docs do not mention that, in order for Traffic Monitor to successfully poll servers using `stats_over_http` output, the experimental `system_stats` ATS plugin must also been enabled.

Otherwise, you get an unhelpful error like `REPORTED - Error parsing loadavg for cache 'edge': Data was missing 'plugin.system_stats.loadavg.one'; eth0: not found in polled data`.

So, this is a good time to make sure our documentation includes that info.

ezelkow1 commented:
> Our docs do not mention that, in order for Traffic Monitor to successfully poll servers using `stats_over_http` output, the experimental `system_stats` ATS plugin must also been enabled.
> 
> Otherwise, you get an unhelpful error like `REPORTED - Error parsing loadavg for cache 'edge': Data was missing 'plugin.system_stats.loadavg.one'; eth0: not found in polled data`.
> 
> So, this is a good time to make sure our documentation includes that info.

Thanks

Updated changelog, removed nil check since it went in to the codebase before this went through, and put info in the docs about adding system_stats

zrhoffman reviewed (approved):
Looks good!
- TM successfully reports cache health when when accepting text/csv polling responses. See the *CDN-in-a-Box CI* result from https://github.com/zrhoffman/trafficcontrol/commits/tm-csv-stats, which uses ATS 9 and stats_over_http with CSV polling
- Code, documentation, and changelog look good
- Tests and benchmarks pass and have good coverage
